// utility/utility.js
var utility = module.exports = {
    defineProperties: function(obj, methods, options){
        var c = {}, item;
        for(var i in methods){
            item = c[i] = {value: methods[i]};
            if(options){
                for(var j in options[i]){
                    item[j] = options[i][j];
                }
            }
        }
        Object.defineProperties(obj, c);
    },
    
    peek: function(val){
        return val[val.length - 1];
    },
    
    date: {
        format: function(source, pattern){
            var o = { 
                'M+': source.getMonth() + 1, //month 
                'd+': source.getDate(), //day 
                'h+': source.getHours(), //hour 
                'm+': source.getMinutes(), //minute 
                's+': source.getSeconds(), //second 
                'q+': Math.floor((source.getMonth() + 3)/3), //quarter 
                'S': source.getMilliseconds() //millisecond 
            }
            if(/(y+)/.test(pattern)) { 
                pattern = pattern.replace(RegExp.$1, (source.getFullYear()+"").substr(4 - RegExp.$1.length)); 
            } 
            
            for(var k in o) {
                if(new RegExp('('+ k +')').test(pattern)) { 
                    pattern = pattern.replace(RegExp.$1, RegExp.$1.length === 1 ? o[k] : ('00' + o[k]).substr(('' + o[k]).length)); 
                } 
            } 
            return pattern;
        },
        
        toMinute: function(msd){//将毫秒转化为xx分钟xx秒的格式
            var second = parseFloat(msd) / 1000,
                sign = second < 0 ? '- ' : '',
                m, s;
            if(!second){return '0:0';}
            second = Math.abs(second);
            m = second >= 60 ? parseInt(second / 60.0) : 0;
            s = parseInt(second >= 60 ? (parseFloat(second / 60.0) - m) * 60 : second);
            return sign + m + ':' + s;
        },
            
        toMinuteString: function(val){//将xx.x或是xx:xx转化为 xx分钟xx秒
            var minute = new String(val), last;
            if(~minute.indexOf(':')){return minute;}
            minute = minute.split('.');
            last = minute.length > 1 ? parseInt(parseFloat('.' + utility.peek(minute)) * 60) : 0;
            return minute[0] + ':' + last;
        },
        
        minuteToMillisecond: function(minute){//将分钟转化为毫秒
            var second = 0;
            if(typeof(minute) === 'string'
                && ~minute.indexOf(':')){
                second = minute.split(':');
                second = second[0] * 60 + parseInt(utility.peek(second));
            }else{
                second = minute * 60;
            }
            return parseInt(second * 1000);
        }
    }
};
